domain/user/state/state.go

```diff
@@ -0,0 +1,248 @@
+			domainUser.Name,
+			domainUser.DisplayName,
+			domainUser.CreatorUUID,
+			domainUser.CreatedAt,
+		)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
+		if databaseutils.IsErrConstraintUnique(err) {
+			return fmt.Errorf("user %q already exists: %w", domainUser.Name, err)
+		}
+		if databaseutils.IsErrConstraintForeignKey(err) {
+			return fmt.Errorf("user %q creator %q does not exist: %w", domainUser.Name, domainUser.CreatorUUID, err)
+		}
+		return fmt.Errorf("cannot insert user %q: %w", domainUser.Name, err)
+	}
+
+	return nil
+}
+
+// AddUserWithPasswordHash will add a new user to the database with the
+// provided password hash and salt. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. If the users creator
+// does not exist or has been previously removed an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUserWithPasswordHash(ctx context.Context, domainUser domainuser.User, passwordHash string, salt []byte) error {
+	err := st.AddUser(ctx, domainUser)
+	if err != nil {
+		return fmt.Errorf("cannot add user %q: %w", domainUser.Name, err)
+	}
+
+	err = st.SetPasswordHash(ctx, domainUser.UUID, passwordHash, salt)
+	if err != nil {
+		return fmt.Errorf("cannot set password hash for user %q: %w", domainUser.Name, err)
+	}
+
+	return nil
+}
+
+// AddUserWithActivationKey will add a new user to the database with the
+// provided activation key. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. if the users creator
+// does not exist or has been previously removed an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUserWithActivationKey(ctx context.Context, user domainuser.User, activationKey string) error {
+	return nil
+}
+
+// GetUser will retrieve the user specified by name from the database where
+// the user is active and has not been removed. If the user does not exist
+// or is deleted an error that satisfies usererrors.NotFound will be
+// returned.
+func (st *State) GetUser(ctx context.Context, userUUID string) (user.User, error) {
+	db, err := st.DB()
+	if err != nil {
+		return user.User{}, fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	getUserQuery := `
+SELECT name, display_name, created_by_uuid, created_at
+FROM user
+WHERE uuid = ? AND removed = false
+`
+
+	var u user.User
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		row := tx.QueryRowContext(ctx, getUserQuery, userUUID)
+
+		var name, displayName, creatorUUID string
+		var createdAt time.Time
+		err = row.Scan(&name, &displayName, &creatorUUID, &createdAt)
+		if err != nil {
+			return fmt.Errorf("cannot scan user %q from DB: %w", userUUID, err)
+		}
+		u = user.User{
+			Name:        name,
+			DisplayName: displayName,
+			CreatorUUID: creatorUUID,
+			CreatedAt:   createdAt,
+		}
+
+		return nil
+	})
+
```

Comment:
You can perform you database error checking down here instead of the transaction.

For example:

```go
if errors.Is(err, sql. ErrNoRows) { return usererrors.NotFound } // Need to annotate more just an example
```